prisma/schema.prisma

```diff
@@ -0,0 +1,126 @@
+// This is your Prisma schema file,
+// learn more about it in the docs: https://pris.ly/d/prisma-schema
+
+generator client {
+  provider = "prisma-client-js"
+}
+
+datasource db {
+  provider = "postgresql"
+  url      = env("DATABASE_URL")
+}
+
+// ================ MODELS ================
+model User {
+  uid String @id
+
+  updatedAt DateTime @updatedAt @map("updated_at")
+
+  applications Application[]
+
+  @@map("users")
+}
+
+model Company {
+  id         Int     @id @default(autoincrement())
+  name       String
+  companyUrl String  @map("company_url")
+  isVerified Boolean @default(false) @map("is_verified")
+
+  updatedAt DateTime @updatedAt @map("updated_at")
+
+  roles Role[]
+
+  @@map("companies")
+}
+
+enum RoleType {
+  WINTER_INTERNSHIP
+  SPRING_INTERNSHIP
+  SUMMER_INTERNSHIP
+  FALL_INTERNSHIP
+  FULL_TIME
+}
+
+model Role {
+  id         Int      @id @default(autoincrement())
+  company    Company  @relation(fields: [companyId], references: [id])
+  companyId  Int      @map("company_id")
+  title      String
+  type       RoleType
+  year       Int
+  isVerified Boolean  @default(false) @map("is_verified")
+
+  updatedAt DateTime @updatedAt @map("updated_at")
+
+  applications Application[]
+
+  @@map("roles")
+}
+
+model Application {
+  id     Int    @id @default(autoincrement())
+  user   User   @relation(fields: [userId], references: [uid])
+  userId String @map("user_id")
+  role   Role   @relation(fields: [roleId], references: [id])
+  roleId Int    @map("role_id")
+
+  updatedAt DateTime @updatedAt @map("updated_at")
+
+  applicationStages ApplicationStage[]
+  tasks             Task[]
+
+  @@unique([userId, roleId])
+  @@map("applications")
+}
+
+enum ApplicationStageType {
+  APPLIED
+  ONLINE_ASSESSMENT
+  TECHNICAL
+  NON_TECHNICAL
+  MIXED
+  OFFERED
+  ACCEPTED
+  REJECTED
+  WITHDRAWN
+}
+
+model ApplicationStage {
+  id            Int                  @id @default(autoincrement())
+  application   Application          @relation(fields: [applicationId], references: [id])
+  applicationId Int                  @map("application_id")
+  type          ApplicationStageType
+  date          DateTime
+  reaction      Reaction?            @relation(fields: [reactionId], references: [id])
+  reactionId    Int?                 @unique @map("reaction_id")
+
+  updatedAt DateTime @updatedAt @map("updated_at")
+
+  @@map("application_stages")
+}
+
+model Reaction {
+  id               Int               @id @default(autoincrement())
+  emojiUnicodeHex  String            @map("emoji_unicode_hex")
+  remark           String
+  applicationStage ApplicationStage?
```

Comment:
Another minor point: if the mapping is `ApplicationStage (1) <-> (0 or 1) Reaction` shouldn't the applicationStage not be nullable or is this a limitation of how Prisma models work?

```suggestion
  applicationStage ApplicationStage
```

Reply:
Yup, this is a restriction imposed by Prisma. Since we do not define the `applicationStageId` on the `Reaction` model, it has to optional.

See: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/one-to-one-relations#required-and-optional-1-1-relation-fields